### PR TITLE
Add tests take with info bad arguments

### PR DESCRIPTION
**Pull request by Lobotuerk**
Signed-off-by: lobotuerk <jtlorente@ekumenlabs.com>

### CODE REVIEW

File: test_rmw_implementation/test/test_subscription.cpp [hunk 1]
```diff
@@ -377,3 +377,34 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), count_mismatched_subs
   EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
   EXPECT_EQ(0u, publisher_count);
 }
+
+TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), take_with_info_with_bad_args) {
+  rmw_ret_t ret;
```
Comment (Blast545) [inline test_rmw_implementation/test/test_subscription.cpp:382]: Suggestion: move to line 388 where ret is used for the first time

File: test_rmw_implementation/test/test_subscription.cpp [hunk 2]
```diff
@@ -377,3 +377,34 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), count_mismatched_subs
...
+  bool taken;
```
Comment (Blast545) [inline test_rmw_implementation/test/test_subscription.cpp:383]: Suggestion: move closer to where taken is used for the first time, and initialize the taken variable.

**Review by Blast545 — APPROVED** [review]
Minor comments, LGTM

File: test_rmw_implementation/test/test_subscription.cpp (hunk 1 above)
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp:382]: But all 5 variables get used on the same line (388), moving one closer would get the other further

Comment (hidmic) [reply test_rmw_implementation/test/test_subscription.cpp:382]: Why is that a problem?

Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp:382]: The problem is that @Blast545 pointed out 2 variables, the 2 further, to move them closer. Whenever I move those 2 closer, the other 2 are going to get moved away from the first place they are used, and the problem will just loop around, as there are 5 variables that get used for the first time in the same place. My question would be, are some variables more important to be closer than the others?

Comment (Blast545) [reply test_rmw_implementation/test/test_subscription.cpp:382]: The first variable, `rmw_ret_t ret` can be initialized in the same place where it is used, by moving it there you are saving a declaration.

As for the second one, `bool taken` yeah, you are right and there is no need to move it. But add an initialization value to its declaration

Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp:382]: Updated

File: test_rmw_implementation/test/test_subscription.cpp [hunk 3]
```diff
@@ -377,3 +377,34 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), count_mismatched_subs
...
+  test_msgs__msg__BasicTypes output_message{};
+  rmw_message_info_t message_info{};
+  rmw_subscription_allocation_t * null_allocation{nullptr}; // still valid allocator
```
[the remark this answers is not in the document]
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp:386]: Updated

File: test_rmw_implementation/test/test_subscription.cpp [hunk 4]
```diff
@@ -377,3 +377,34 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), count_mismatched_subs
   EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
   EXPECT_EQ(0u, publisher_count);
 }
+
+TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), take_with_info_with_bad_args) {
+  rmw_ret_t ret;
+  bool taken;
+  test_msgs__msg__BasicTypes output_message{};
+  rmw_message_info_t message_info{};
```
[the remark this answers is not in the document]
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp:385]: Updated

File: test_rmw_implementation/test/test_subscription.cpp (hunk 2 above)
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp]: updated

File: test_rmw_implementation/test/test_subscription.cpp [hunk 5]
```diff
@@ -377,3 +377,34 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), count_mismatched_subs
...
+  rmw_subscription_allocation_t * null_allocation{nullptr}; // still valid allocator
+
+  ret = rmw_take_with_info(nullptr, &output_message, &taken, &message_info, null_allocation);
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, ret) << rmw_get_error_string().str;
+  rmw_reset_error();
```
[the remark this answers is not in the document]
Comment (Lobotuerk) [reply test_rmw_implementation/test/test_subscription.cpp:402]: updated

**Comment by Lobotuerk** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12396/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7364/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10107/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12309/)

File: test_rmw_implementation/test/test_subscription.cpp [hunk 6]
```diff
@@ -23,17 +23,20 @@
...
+#include "rmw_dds_common/gid_utils.hpp"
+
 #ifdef RMW_IMPLEMENTATION
 # define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
 # define CLASSNAME(NAME, SUFFIX) CLASSNAME_(NAME, SUFFIX)
 #else
 # define CLASSNAME(NAME, SUFFIX) NAME
 #endif
 
+using rmw_dds_common::operator==;
```
Comment (hidmic) [inline test_rmw_implementation/test/test_subscription.cpp:38]: @Lobotuerk nit: add a note explaining why this line's here.

**Review by hidmic — APPROVED** [review]

**Comment by Lobotuerk** [comment]
CI job up to `test_rmw_implementation` after rebase:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12433/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7400/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10143/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12346/)

**Comment by Lobotuerk** [comment]
CI with windows fixes

**Comment by Lobotuerk** [comment]
All CI green, merging

**Pull request merged**